### PR TITLE
Document external libraries and namespaces

### LINKED ISSUE #49: Document external library support and namespaces
References:
  * https://github.com/Zilliqa/scilla/pull/478
  * https://github.com/Zilliqa/scilla/pull/507
  * https://github.com/Zilliqa/scilla/pull/526

### PR DESCRIPTION
**Pull request by vaivaswatha**
Closes #49. 

References:

Zilliqa/scilla#478
Zilliqa/scilla#507
Zilliqa/scilla#526

### CODE REVIEW

File: docs/source/scilla-in-depth.rst [hunk 1]
```diff
@@ -2093,10 +2093,123 @@ PairUtils
...
+              "arguments" : ["ExampleLib", "0x986556789012345678901234567890123456abcd"]
+          },
+          ...
+      ]
+    }
+  ]
+
+Namespaces
+**********
+When a contract or a library imports more than one library, a possibility of name clash
```
Comment (jjcnn) [inline docs/source/scilla-in-depth.rst]: How about this instead:

"Import statements can be used to define separate namespaces for imported names. To push the names from a library ``Foo`` into the namespace ``Bar``, use the statement ``import Foo as Bar``. Accessing a variable ``v`` in ``Foo`` must now be done using the qualified name ``Bar.v``. This is useful when importing multiple libraries that define the same name.

The same variable name must not occur in the same namespace, so if multiple imported libraries define the same name, then at most one of the libraries may reside in the default (unqualified) namespace. All other conflicting libraries must be pushed to separate namespaces.

Extending our previous example, ..."

File: docs/source/scilla-in-depth.rst [hunk 2]
```diff
@@ -2093,10 +2093,123 @@ PairUtils
...
+with ``Bar.``.
+
+Extending our previous example, shown below is a contract that imports ``ExampleLib``
+in namespace ``Bar``, to use the function ``add_if_equal``.
+
+.. code-block:: ocaml
+
+  scilla_version 0
+
+  import ExampleLib as Foo
```
Comment (jjcnn) [inline docs/source/scilla-in-depth.rst]: The text says we are importing the library into the namespace ``Bar``.

File: docs/source/scilla-in-depth.rst [hunk 3]
```diff
@@ -2093,10 +2093,123 @@ PairUtils
...
+turn imports library ``Z``, then the names in ``Z`` are not in scope in `X``, but only
+in ``Y``. Cyclic dependencies in imports are not allowed and flagged as errors
+during the checking phase.
+
+
+Local Development with User Defined Libraries
+*********************************************
+
+To use library functions in an external (user-defined) library module, the command line
+argument to the scilla executables must include a ``-libdir`` option, along with a list of
```
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply docs/source/scilla-in-depth.rst]: The referenced link there actually has the details on the command line interface with examples, so do we want one here?

[the remark this answers is not in the document]
Comment (jjcnn) [reply docs/source/scilla-in-depth.rst]: Ah, you're right. I didn't see the reference.

File: docs/source/scilla-in-depth.rst [hunk 4]
```diff
@@ -2174,37 +2178,35 @@ in ``Y``. Cyclic dependencies in imports are not allowed and flagged as errors
...
+User-defined Libraries on the Blockchain
 ****************************************
 
 While the Zilliqa blockchain is designed to provide the standard Scilla libraries to an
-executing contract, it must be provided with extra information to support user defined
+executing contract, it must be provided with extra information to support user-defined
 libraries.
 
-Contracts (or libraries) that import user defined libraries must include, in its `init.json`
+A contracts or a library that import user-defined libraries must include in its `init.json`
```
Comment (jjcnn) [inline docs/source/scilla-in-depth.rst]: "import" -> "imports"

File: docs/source/scilla-in-depth.rst [hunk 5]
```diff
@@ -2226,15 +2228,14 @@ init.json should have an entry:
...
+Import statements can be used to define separate namespaces for imported names.
+To push the names from a library ``Foo`` into the namespace ``Bar``, use the statement
+``import Foo as Bar``. Accessing a variable ``v`` in Foo must now be done using the qualified
+name ``Bar.v``. This is useful when importing multiple libraries that define the same name.
 
-To enable users to meaningfully import libraries that may have name clashes, Scilla provides
-an ``import as`` construct. ``import Foo as Bar`` imports the library ``Foo`` with all of its
-names pushed into the ``Bar`` namespace. Accesses to names in ``Foo`` must now be prefixed
-with ``Bar.``.
+The same variable name must not occur in the same namespace, so if multiple imported libraries
```
Comment (jjcnn) [inline docs/source/scilla-in-depth.rst]: "The same variable name must not occur in the same namespace,..." ->
"The same variable name must not be defined more than once in the same namespace, ..."

I'm surprised neither of us caught the nonsense of that sentence. :-)

**Comment by vaivaswatha** [comment]
> A couple of minor issues still remain (one of them, sadly, due to me...).
> 
> What do we do in terms of approving the PR? I don't really like to approve it, because they someone might merge it by mistake before the blockchain supports external libraries. Should we mark the PR as a draft?

Sorry I forgot to respond to this. You are right. I will mark it as a draft.

**Comment by vaivaswatha** [comment]
> > A couple of minor issues still remain (one of them, sadly, due to me...).
> > What do we do in terms of approving the PR? I don't really like to approve it, because they someone might merge it by mistake before the blockchain supports external libraries. Should we mark the PR as a draft?
> 
> Sorry I forgot to respond to this. You are right. I will mark it as a draft.

Ah, looks like I can't do that. https://github.community/t5/How-to-use-Git-and-GitHub/Feature-Request-Switch-from-ready-to-draft-in-pull-requests/td-p/19107.

Let's just leave this as it is without you approving it. I'll fix the remaining minor issues and when the blockchain implementation is present, will ping you again to approve it.

File: docs/source/scilla-in-depth.rst (hunk 5 above)
Comment (vaivaswatha) [reply docs/source/scilla-in-depth.rst]: ha ha, you're right.

**Comment by jjcnn** [comment]
> > I will mark it as a draft.
> 
> Ah, looks like I can't do that. https://github.community/t5/How-to-use-Git-and-GitHub/Feature-Request-Switch-from-ready-to-draft-in-pull-requests/td-p/19107.
> 
> Let's just leave this as it is without you approving it. I'll fix the remaining minor issues and when the blockchain implementation is present, will ping you again to approve it.

1 penalty point for GitHub. :-(

My review is still listed as "Changes requested", which is now incorrect. :-)

**Comment by vaivaswatha** [comment]
Merging this since the [blockchain support](https://github.com/Zilliqa/Zilliqa/pull/1981) is now merged.

**Pull request merged**